packages/go_router/lib/src/delegate.dart

```diff
@@ -42,22 +43,57 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
   /// Set to true to disable creating history entries on the web.
   final bool routerNeglect;
 
-  final GlobalKey<NavigatorState> _key = GlobalKey<NavigatorState>();
-
-  RouteMatchList _matches = RouteMatchList.empty();
+  RouteMatchList _matchList = RouteMatchList.empty();
   final Map<String, int> _pushCounts = <String, int>{};
+  final RouteConfiguration _configuration;
+
+  @override
+  Future<bool> popRoute() async {
+    // Iterate backwards through the RouteMatchList until seeing a GoRoute
+    // with a non-null parentNavigatorKey or a ShellRoute with a non-null parentNavigatorKey
+    // and pop from that Navigator instead of the root.
+    final int matchCount = _matchList.matches.length;
+    for (int i = matchCount - 1; i >= 0; i--) {
+      final RouteMatch match = _matchList.matches[i];
+      final RouteBase route = match.route;
+
+      // If this is a ShellRoute, then pop one of the subsequent GoRoutes, if
+      // there are any.
+      if (route is ShellRoute && (matchCount - i) > 2) {
+        final NavigatorState? navigator = route.navigatorKey.currentState;
+        final bool didPop = await navigator?.maybePop() ?? false;
+        if (didPop) {
+          return didPop;
+        }
+      }
```

Comment:
This should also take care of GoRoute whose parentNavigatorKey is not null

Reply:
In that case, I think we will pop from the root Navigator or ShellRoute Navigator, or are you saying there's a bug here?

Reply:
Are you sure it's safe to assert false? What if the user is using WillPopScope?

Reply:
WillPopScope will not make it return false, it will only make it return true if it handles it or let the navigator handle the pop. and navigator will only return false if it only has one page which shouldn't be possible.

Reply:
Fixed in d35a25b736d3b370953a11e10a7457fba8005b95